src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -0,0 +1,551 @@
+
+    return data_dir
+
+
+def _get_assembly_name_upa_mapping(conf: Conf, workspace_id: int) -> dict[str, str]:
+    """
+    Get a mapping of UPA to assembly name from target workspace.
+    """
+    assembly_objs = loader_helper.list_objects(
+        workspace_id, conf, loader_common_names.OBJECTS_NAME_ASSEMBLY
+    )
+    hashmap = {obj_info[1]: "{6}_{0}_{4}".format(*obj_info) for obj_info in assembly_objs}
+    return hashmap
+ 
+
+def create_entries_in_sourcedata_workspace(
+        root_dir: str,
+        env: str,
+        workspace_id: int,
+        assembly_names: list[str],
+        assembly_name_to_dir: dict[str, str],
+        assembly_name_to_upa: dict[str, str],
+        output_dir: str,
+) -> list[str]:
+    """
+    Create a standard entry in sourcedata/workspace for each assembly.
+    Hardlink to the original assembly file in sourcedata to avoid duplicating the file.
+    Update the uploaded.yaml file in the genome directory with assembly names and upa info.
+
+    root_dir: root directory
+    env: KBase environment
+    workspace_id: target workspace addressed by the permanent ID
+    assembly_names: a list of assembly names newly uploaded to the target workspace
+    assembly_name_to_dir: a dictionary of assembly name to its directory path
+    assembly_name_to_upa: a dictionary of assembly name to its UPA
+    output_dir: output directory to create entries in workspace
+    """
+    upas = list()
+    for assembly_name in assembly_names:
+        try:
+            assembly_dir = assembly_name_to_dir[assembly_name]
+        except KeyError as e:
+            raise ValueError(f"Unable to find assembly {assembly_name}") from e
+
+        src_file = _get_source_file(assembly_dir, assembly_name)
+        assembly_dir_name = os.path.basename(assembly_dir)
+
+        try:
+            upa = assembly_name_to_upa[assembly_name]
+        except KeyError as e:
+            raise ValueError(f"Unable to find assembly {assembly_name} from target workspace") from e
+        
+        upa_dir = os.path.join(output_dir, upa)
+        os.makedirs(upa_dir, exist_ok=True)
+        upas.append(upa)
+
+        dest_file = os.path.join(upa_dir, assembly_name)
+        loader_helper.create_hardlink_between_files(dest_file, src_file)
+        _update_yaml_file(root_dir, env, workspace_id, upa, assembly_dir_name, assembly_name)
+
+    return upas
+
+
+def upload_assemblies_to_workspace(
+        conf: Conf,
+        workspace_name: str,
+        data_dir: str,
+) -> list[str]:
+    """
+    Upload assemblies to workspace and record failed assembly names.
+
+    conf: Conf object
+    workspace_name: a string used as a name for a workspace
+    data_dir: directory of assemblies to upload
+    """
+
+    assembly_files = os.listdir(data_dir)
+    print(f'start uploading {len(assembly_files)} assembly files')
+
+    failed_names = list()
+
+    counter = 1
+    for assembly_name in assembly_files:
```

Comment:
This is entirely serial - to get reasonable upload speeds it needs to be parallelized like the downloader

Reply:
I think part of the issue is that the Conf code is C&P from the downloader, rather than moved into a shared module and reused

Reply:
@MrCreosote 
I am trying to parallelize the uploader. My current implementation passes in conf object as part of batch_input, which causes the 'pickle' issue in multiprocessing. Do you have any suggestions on this? (multiprocess manager, other ways to tackle this)

Reply:
How does the downloader get around the problem?

Reply:
Updated!

Reply:
:+1: